# Data Cleaning And Visualisation Task

## Loading in the data:
We know that the json data provided to us is in the form:
 
`{data: [{}, {}, {}, …, {}]}`

So the first thing to do is to load the data in. This will be done using the `json` module in python

In [1]:
import pandas as pd
import json

fh = open('../data/tasktwo/data.json')
alldata = json.load(fh)
fh.close()
alldata['data'][0]

{'category': 'Sports',
 'event_name': 'View Project',
 'gender': 'M',
 'age': '18-24',
 'marital_status': 'married',
 'session_id': '69f62d2ae87640f5a2dde2b2e9229fe6',
 'device': 'android',
 'client_time': 1393632004,
 'location': {'latitude': 40.189788,
  'city': 'Lyons',
  'state': 'CO',
  'longitude': -105.35528,
  'zip_code': '80540'}}

We can see what the first object in our data looks like above

We have also imported the `pandas` module because we plan to convert the json data into a dataframe to process it further.

In [2]:
df = pd.DataFrame(alldata['data'])
df

category    event_name gender    age marital_status  \
0           Sports  View Project      M  18-24        married   
1       Technology  View Project      M  18-24         single   
2      Environment  View Project      M    55+         single   
3       Technology  View Project      M  18-24         single   
4           Sports  View Project      F  25-34        married   
...            ...           ...    ...    ...            ...   
49995       Sports  Fund Project      F  18-24        married   
49996   Technology  View Project      F  45-54         single   
49997      Fashion  View Project      F  25-34         single   
49998       Sports  Fund Project      F  35-44        married   
49999      Fashion  View Project      M  18-24         single   

                             session_id   device  client_time  \
0      69f62d2ae87640f5a2dde2b2e9229fe6  android   1393632004   
1      4459d001feb8438eae5f4ec24abcd992      iOS   1393632022   
2      0db9ed700a184d48a9d04806696e3642      iOS   1393632032   
3      68195e2372bd4022b17220fc21de9138  android   1393632038   
4      9508a8385dc94773baba8aa7d1c2aa75      iOS   1393632051   
...                                 ...      ...          ...   
49995  412b973788704c36a008a506cdbba033      iOS   1396309720   
49996  4addd36a6f4347c59865fe04a92e8d57  android   1396309769   
49997  3181a138a0b94d1da9a3d29f7816fcc7      iOS   1396309893   
49998  2df6b9a0c66b48a389330327517b9276      iOS   1396309968   
49999  fc24339bfda84ff29102981cbee78023      iOS   1396310325   

                                                location  amount  
0      {'latitude': 40.189788, 'city': 'Lyons', 'stat...     NaN  
1      {'latitude': 33.844371, 'city': 'Alpharetta', ...     NaN  
2      {'latitude': 42.446396, 'city': 'Westford', 's...     NaN  
3      {'latitude': 44.624413, 'city': 'Saranac', 'st...     NaN  
4      {'latitude': 36.747083, 'city': 'Lampe', 'stat...     NaN  
...                                                  ...     ...  
49995  {'latitude': 45.343615, 'city': 'West Linn', '...    35.0  
49996  {'latitude': 41.079983, 'city': 'Greenwich', '...     NaN  
49997  {'latitude': 40.36502, 'city': 'Irwin', 'state...     NaN  
49998  {'latitude': 39.477625, 'city': 'Martinsville'...    37.0  
49999  {'latitude': 45.408374, 'city': 'Lake Oswego',...     NaN  

[50000 rows x 10 columns]

## Cleaning the data
The first criteria provided in the problem statement is that the bicycle project belongs to two categories
- Sports
- Environment

So we can safely assume that the analytics data for any other category is redundant for our use case, and hence we can get rid of it.

To do this, we create a boolean mask in pandas and apply it to our dataframe `df`

In [3]:
mask = (df['category'] == "Sports")|(df['category'] == "Environment")
df = df[mask]
df

category    event_name gender    age marital_status  \
0           Sports  View Project      M  18-24        married   
2      Environment  View Project      M    55+         single   
4           Sports  View Project      F  25-34        married   
12          Sports  View Project      F  18-24        married   
16          Sports  Fund Project      M  18-24        married   
...            ...           ...    ...    ...            ...   
49984  Environment  Fund Project      M  18-24        married   
49989       Sports  Fund Project      M  18-24         single   
49994  Environment  Fund Project      F  35-44         single   
49995       Sports  Fund Project      F  18-24        married   
49998       Sports  Fund Project      F  35-44        married   

                             session_id   device  client_time  \
0      69f62d2ae87640f5a2dde2b2e9229fe6  android   1393632004   
2      0db9ed700a184d48a9d04806696e3642      iOS   1393632032   
4      9508a8385dc94773baba8aa7d1c2aa75      iOS   1393632051   
12     a1fdfe0bebed4510a9059bcfb3ba1325      iOS   1393632171   
16     69f62d2ae87640f5a2dde2b2e9229fe6  android   1393632200   
...                                 ...      ...          ...   
49984  584713f5385a4ce2b881e877f1fa9a80  android   1396308722   
49989  4829aa471c6b431fa023076746f7476f      iOS   1396309022   
49994  dbe304febcd94d768740237f37e918cd  android   1396309700   
49995  412b973788704c36a008a506cdbba033      iOS   1396309720   
49998  2df6b9a0c66b48a389330327517b9276      iOS   1396309968   

                                                location  amount  
0      {'latitude': 40.189788, 'city': 'Lyons', 'stat...     NaN  
2      {'latitude': 42.446396, 'city': 'Westford', 's...     NaN  
4      {'latitude': 36.747083, 'city': 'Lampe', 'stat...     NaN  
12     {'latitude': 33.794055, 'city': 'Atlanta', 'st...     NaN  
16     {'latitude': 40.189788, 'city': 'Lyons', 'stat...    31.0  
...                                                  ...     ...  
49984  {'latitude': 33.760505, 'city': 'Atlanta', 'st...    23.0  
49989  {'latitude': 39.872535, 'city': 'Westminster',...    37.0  
49994  {'latitude': 27.457947, 'city': 'Sebring', 'st...    35.0  
49995  {'latitude': 45.343615, 'city': 'West Linn', '...    35.0  
49998  {'latitude': 39.477625, 'city': 'Martinsville'...    37.0  

[20079 rows x 10 columns]

We can see now that we have successfully narrowed down our dataframe to 20079 rows from 50000 rows!

Our next step is going to be to try and drop the `event_name` column.

In [4]:
masktemp = df['amount'] == 0
masktemp.sum() #Sum of bool mask is num of trues in the mask

0

As we can see there are no people who have funded a project with amount = 0.

So we can treat the people who have only viewed the project as people who have funded the project with amount = 0, and drop the `event_name` column altogether.

In [5]:
df['amount'] = df['amount'].fillna(0)
df = df.drop(columns=['event_name'])
df

category gender    age marital_status  \
0           Sports      M  18-24        married   
2      Environment      M    55+         single   
4           Sports      F  25-34        married   
12          Sports      F  18-24        married   
16          Sports      M  18-24        married   
...            ...    ...    ...            ...   
49984  Environment      M  18-24        married   
49989       Sports      M  18-24         single   
49994  Environment      F  35-44         single   
49995       Sports      F  18-24        married   
49998       Sports      F  35-44        married   

                             session_id   device  client_time  \
0      69f62d2ae87640f5a2dde2b2e9229fe6  android   1393632004   
2      0db9ed700a184d48a9d04806696e3642      iOS   1393632032   
4      9508a8385dc94773baba8aa7d1c2aa75      iOS   1393632051   
12     a1fdfe0bebed4510a9059bcfb3ba1325      iOS   1393632171   
16     69f62d2ae87640f5a2dde2b2e9229fe6  android   1393632200   
...                                 ...      ...          ...   
49984  584713f5385a4ce2b881e877f1fa9a80  android   1396308722   
49989  4829aa471c6b431fa023076746f7476f      iOS   1396309022   
49994  dbe304febcd94d768740237f37e918cd  android   1396309700   
49995  412b973788704c36a008a506cdbba033      iOS   1396309720   
49998  2df6b9a0c66b48a389330327517b9276      iOS   1396309968   

                                                location  amount  
0      {'latitude': 40.189788, 'city': 'Lyons', 'stat...     0.0  
2      {'latitude': 42.446396, 'city': 'Westford', 's...     0.0  
4      {'latitude': 36.747083, 'city': 'Lampe', 'stat...     0.0  
12     {'latitude': 33.794055, 'city': 'Atlanta', 'st...     0.0  
16     {'latitude': 40.189788, 'city': 'Lyons', 'stat...    31.0  
...                                                  ...     ...  
49984  {'latitude': 33.760505, 'city': 'Atlanta', 'st...    23.0  
49989  {'latitude': 39.872535, 'city': 'Westminster',...    37.0  
49994  {'latitude': 27.457947, 'city': 'Sebring', 'st...    35.0  
49995  {'latitude': 45.343615, 'city': 'West Linn', '...    35.0  
49998  {'latitude': 39.477625, 'city': 'Martinsville'...    37.0  

[20079 rows x 9 columns]

We can also consider dropping the `category` column as well, since both categories are of equal interest to us.

In [6]:
df = df.drop(columns=['category'])
df

gender    age marital_status                        session_id   device  \
0          M  18-24        married  69f62d2ae87640f5a2dde2b2e9229fe6  android   
2          M    55+         single  0db9ed700a184d48a9d04806696e3642      iOS   
4          F  25-34        married  9508a8385dc94773baba8aa7d1c2aa75      iOS   
12         F  18-24        married  a1fdfe0bebed4510a9059bcfb3ba1325      iOS   
16         M  18-24        married  69f62d2ae87640f5a2dde2b2e9229fe6  android   
...      ...    ...            ...                               ...      ...   
49984      M  18-24        married  584713f5385a4ce2b881e877f1fa9a80  android   
49989      M  18-24         single  4829aa471c6b431fa023076746f7476f      iOS   
49994      F  35-44         single  dbe304febcd94d768740237f37e918cd  android   
49995      F  18-24        married  412b973788704c36a008a506cdbba033      iOS   
49998      F  35-44        married  2df6b9a0c66b48a389330327517b9276      iOS   

       client_time                                           location  amount  
0       1393632004  {'latitude': 40.189788, 'city': 'Lyons', 'stat...     0.0  
2       1393632032  {'latitude': 42.446396, 'city': 'Westford', 's...     0.0  
4       1393632051  {'latitude': 36.747083, 'city': 'Lampe', 'stat...     0.0  
12      1393632171  {'latitude': 33.794055, 'city': 'Atlanta', 'st...     0.0  
16      1393632200  {'latitude': 40.189788, 'city': 'Lyons', 'stat...    31.0  
...            ...                                                ...     ...  
49984   1396308722  {'latitude': 33.760505, 'city': 'Atlanta', 'st...    23.0  
49989   1396309022  {'latitude': 39.872535, 'city': 'Westminster',...    37.0  
49994   1396309700  {'latitude': 27.457947, 'city': 'Sebring', 'st...    35.0  
49995   1396309720  {'latitude': 45.343615, 'city': 'West Linn', '...    35.0  
49998   1396309968  {'latitude': 39.477625, 'city': 'Martinsville'...    37.0  

[20079 rows x 8 columns]

Now if we take a closer look, we observe that one user has made multiple entries in the dataframe, and each of these entries is differentiated by its timestamp. This can be easily shown by executing the cell below:

In [7]:
df.duplicated(['session_id']).sum(), df.duplicated(['session_id', 'client_time']).sum()

(12961, 0)

This is probably a good time to unfold our `location` column. We left it as a dictionary uptil now, but we can't group unhashable types of data. So we will open up the `location` dictionary into multiple columns. 

In [8]:
locdf = pd.json_normalize(df['location']) #Faster as compared to df['location'].apply(pd.Series)
locdf

latitude          city state   longitude zip_code
0      40.189788         Lyons    CO -105.355280    80540
1      42.446396      Westford    MA  -71.459405    01886
2      36.747083         Lampe    MO  -93.458626    65681
3      33.794055       Atlanta    GA  -84.377326    30308
4      40.189788         Lyons    CO -105.355280    80540
...          ...           ...   ...         ...      ...
20074  33.760505       Atlanta    GA  -84.399626    30313
20075  39.872535   Westminster    CO -105.034487    80030
20076  27.457947       Sebring    FL  -81.373012    33870
20077  45.343615     West Linn    OR -122.664926    97068
20078  39.477625  Martinsville    IN  -86.466792    46151

[20079 rows x 5 columns]

In [9]:
locdf = locdf.set_index(df.index)
df = df.join(locdf, on=df.index)
df = df.drop(columns=['location'])
df

gender    age marital_status                        session_id   device  \
0          M  18-24        married  69f62d2ae87640f5a2dde2b2e9229fe6  android   
2          M    55+         single  0db9ed700a184d48a9d04806696e3642      iOS   
4          F  25-34        married  9508a8385dc94773baba8aa7d1c2aa75      iOS   
12         F  18-24        married  a1fdfe0bebed4510a9059bcfb3ba1325      iOS   
16         M  18-24        married  69f62d2ae87640f5a2dde2b2e9229fe6  android   
...      ...    ...            ...                               ...      ...   
49984      M  18-24        married  584713f5385a4ce2b881e877f1fa9a80  android   
49989      M  18-24         single  4829aa471c6b431fa023076746f7476f      iOS   
49994      F  35-44         single  dbe304febcd94d768740237f37e918cd  android   
49995      F  18-24        married  412b973788704c36a008a506cdbba033      iOS   
49998      F  35-44        married  2df6b9a0c66b48a389330327517b9276      iOS   

       client_time  amount   latitude          city state   longitude zip_code  
0       1393632004     0.0  40.189788         Lyons    CO -105.355280    80540  
2       1393632032     0.0  42.446396      Westford    MA  -71.459405    01886  
4       1393632051     0.0  36.747083         Lampe    MO  -93.458626    65681  
12      1393632171     0.0  33.794055       Atlanta    GA  -84.377326    30308  
16      1393632200    31.0  40.189788         Lyons    CO -105.355280    80540  
...            ...     ...        ...           ...   ...         ...      ...  
49984   1396308722    23.0  33.760505       Atlanta    GA  -84.399626    30313  
49989   1396309022    37.0  39.872535   Westminster    CO -105.034487    80030  
49994   1396309700    35.0  27.457947       Sebring    FL  -81.373012    33870  
49995   1396309720    35.0  45.343615     West Linn    OR -122.664926    97068  
49998   1396309968    37.0  39.477625  Martinsville    IN  -86.466792    46151  

[20079 rows x 12 columns]

We can safely infer that the interest of a user in the categories of our project will be directly related to the number of times the user shows up in this dataset.

We can now introduce an additional column to our dataset which will show us "measure of interest" a particular user has in the the categories of our product. We will call it frequency, and it will represent the number of times a particular entry has been made in our dataset

Initially, the frequency will be set to 1 in each row of our data, because each row represents one viewing or funding done by a user. This will also help us when we group the data (as we shall see)

In [10]:
df['frequency'] = 1
df

gender    age marital_status                        session_id   device  \
0          M  18-24        married  69f62d2ae87640f5a2dde2b2e9229fe6  android   
2          M    55+         single  0db9ed700a184d48a9d04806696e3642      iOS   
4          F  25-34        married  9508a8385dc94773baba8aa7d1c2aa75      iOS   
12         F  18-24        married  a1fdfe0bebed4510a9059bcfb3ba1325      iOS   
16         M  18-24        married  69f62d2ae87640f5a2dde2b2e9229fe6  android   
...      ...    ...            ...                               ...      ...   
49984      M  18-24        married  584713f5385a4ce2b881e877f1fa9a80  android   
49989      M  18-24         single  4829aa471c6b431fa023076746f7476f      iOS   
49994      F  35-44         single  dbe304febcd94d768740237f37e918cd  android   
49995      F  18-24        married  412b973788704c36a008a506cdbba033      iOS   
49998      F  35-44        married  2df6b9a0c66b48a389330327517b9276      iOS   

       client_time  amount   latitude          city state   longitude  \
0       1393632004     0.0  40.189788         Lyons    CO -105.355280   
2       1393632032     0.0  42.446396      Westford    MA  -71.459405   
4       1393632051     0.0  36.747083         Lampe    MO  -93.458626   
12      1393632171     0.0  33.794055       Atlanta    GA  -84.377326   
16      1393632200    31.0  40.189788         Lyons    CO -105.355280   
...            ...     ...        ...           ...   ...         ...   
49984   1396308722    23.0  33.760505       Atlanta    GA  -84.399626   
49989   1396309022    37.0  39.872535   Westminster    CO -105.034487   
49994   1396309700    35.0  27.457947       Sebring    FL  -81.373012   
49995   1396309720    35.0  45.343615     West Linn    OR -122.664926   
49998   1396309968    37.0  39.477625  Martinsville    IN  -86.466792   

      zip_code  frequency  
0        80540          1  
2        01886          1  
4        65681          1  
12       30308          1  
16       80540          1  
...        ...        ...  
49984    30313          1  
49989    80030          1  
49994    33870          1  
49995    97068          1  
49998    46151          1  

[20079 rows x 13 columns]

Now we can eliminate further redundancies by grouping all data points (rows) with the same user.

We could not do this before because the number of times a person shows up in our dataset tells us something about his or her interest in projects of this category, but now we are keeping track of the same using the frequency column, so we can get rid of the client_time column and start grouping the data.

Note: We will be taking the sum of the frequencies as well as the amount. Since this way we will keep track of the total amount spent by a user in funding projects similar to ours. 


In [11]:
df = df.drop(columns=['client_time'])

We should first see whether there are any sessions with different data in other column.

In [12]:
temp_df_1 = df.groupby(['session_id', 'age', 'marital_status', 'device', 'gender', 'latitude', 'longitude'], as_index=False)
temp_df_2 = df.groupby(['session_id'], as_index=False)
len(temp_df_1), len(temp_df_2)
#.agg({'frequency': 'sum', 'amount': 'sum', 'city': 'first', 'state': 'first'})

(8276, 7118)

From the above cell, we can conclude that there are around (8276-7118) cells which have the same session_id but at least one different value in some other column.
They could be differing in age or gender etc. So it's best to treat these as different users filling our survey from the same device. For example, a person and his family can use the same desktop to fill out forms.

So we will group by all values that should be same for one user, instead of just the session_id *

In [13]:
df = df.groupby(['session_id', 'age', 'marital_status', 'device', 'gender', 'latitude', 'longitude', 'city', 'state', 'zip_code'], as_index=False).agg({'frequency': 'sum', 'amount': 'sum'})

df

session_id    age marital_status   device gender  \
0     000716c0e76a41508feb786d202748e7  35-44         single      iOS      M   
1     000716c0e76a41508feb786d202748e7  35-44         single      iOS      U   
2     000e13dfa71146e192060542cda8721d  18-24         single  android      F   
3     0018be2c216249469b0f04f51ca9bc0a  25-34        married  android      M   
4     002a9c51650d47f281518d91717c8d31  25-34        married      iOS      F   
...                                ...    ...            ...      ...    ...   
8271  ffc1fdad55a94e9bab68c720621f1f9e  35-44        married      iOS      F   
8272  ffcb41d4b1184ccaa9f6a3f5ba85fc83  18-24        married      iOS      F   
8273  ffd481352bc54b47a613ec2440b38203  25-34        married      iOS      F   
8274  ffdff9e658e245a1b13540b04ca13f5a  45-54        married      iOS      M   
8275  fff2dc58df814daf8cd28cc28a67dbae  35-44        married      iOS      M   

       latitude   longitude         city state zip_code  frequency  amount  
0     31.711209 -106.243004      El Paso    TX    79936          3    37.0  
1     31.711209 -106.243004      El Paso    TX    79936          1     0.0  
2     33.794055  -84.377326      Atlanta    GA    30308          2    29.0  
3     44.955615  -93.167020   Saint Paul    MN    55104          2     0.0  
4     40.354088  -86.118782  Sharpsville    IN    46068          2    33.0  
...         ...         ...          ...   ...      ...        ...     ...  
8271  40.045073  -87.175046    Hillsboro    IN    47949          3    41.0  
8272  45.455090 -122.696724     Portland    OR    97219          1     0.0  
8273  32.856632  -86.652875      Clanton    AL    35045          2    16.0  
8274  39.470986  -87.411070  Terre Haute    IN    47809          2    43.0  
8275  40.951589  -85.854151        Roann    IN    46974          1     0.0  

[8276 rows x 12 columns]

We have now reduced the data to just 8.3k rows!

\* We can see that there are different users that have used the same session_id in the table above. For example, the very first two rows have the same session_id but different genders. (session_id = 000716c0e76a41508feb786d202748e7)

Since we need \$20 of funding for our project, we should use that as a threshold. Meaning, people who have funded a total amount more than \$20 are willing to spend that amount to fund various projects, which may include ours.

Note: Having grouped the data by users, we gained an advantage. Now we can easily check if a person has donated over \$20 in total or not. His/Her \$20 may or may not be on one single project, but that's a different story.

In [14]:
df = df[df['amount']>20]
df

session_id    age marital_status   device gender  \
0     000716c0e76a41508feb786d202748e7  35-44         single      iOS      M   
2     000e13dfa71146e192060542cda8721d  18-24         single  android      F   
4     002a9c51650d47f281518d91717c8d31  25-34        married      iOS      F   
5     003368a440394561a4e34a09a738e29f  18-24        married  android      M   
8     0045a427d12945e8863ce57d5e6ee049  18-24        married  android      F   
...                                ...    ...            ...      ...    ...   
8268  ff9e86d093fd46dd83492b8e552537a3  45-54        married  android      U   
8269  ffb0c38864ee40beb05f094f44a43e32  18-24        married  android      M   
8270  ffbc07891604406b952dc99158c25240    55+        married      iOS      F   
8271  ffc1fdad55a94e9bab68c720621f1f9e  35-44        married      iOS      F   
8274  ffdff9e658e245a1b13540b04ca13f5a  45-54        married      iOS      M   

       latitude   longitude         city state zip_code  frequency  amount  
0     31.711209 -106.243004      El Paso    TX    79936          3    37.0  
2     33.794055  -84.377326      Atlanta    GA    30308          2    29.0  
4     40.354088  -86.118782  Sharpsville    IN    46068          2    33.0  
5     40.427541 -105.383062        Drake    CO    80515          4    48.0  
8     33.563121 -102.026405    Wolfforth    TX    79382          8   195.0  
...         ...         ...          ...   ...      ...        ...     ...  
8268  41.707918  -81.199030  Painesville    OH    44077          2    37.0  
8269  40.370748 -105.323971   Estes Park    CO    80517          2    26.0  
8270  44.867809  -91.943198    Menomonie    WI    54751          3    59.0  
8271  40.045073  -87.175046    Hillsboro    IN    47949          3    41.0  
8274  39.470986  -87.411070  Terre Haute    IN    47809          2    43.0  

[5291 rows x 12 columns]

We now arrive at a staggering 5.2k rows, having reduced our original dataset by almost 90% we can now move on to visualize our data.

Before that however, it's good to sort the data. The obvious key for sorting will be the total amount given by a user in funding.

In [15]:
df.sort_values(by=['amount'], ascending=False)

session_id    age marital_status   device gender  \
3458  6b625dff333d480d9396cdc74f3c7275    55+        married      iOS      M   
4139  7f9c6efbf9964e8095772961442dd2d7  18-24         single  android      M   
1898  3ab197e44fc24289a08572bba665f702  35-44         single  android      F   
4581  8c9a458eafe7490eb7c3657d15e5fc10  18-24        married      iOS      F   
7131  db3131b5dce54211969002b2f554698b  18-24        married      iOS      F   
...                                ...    ...            ...      ...    ...   
7113  da9516dea6404af19e076f27c31007ae  45-54        married      iOS      U   
7148  dbe35288e8b74ece90d917648f80c514  18-24         single  android      U   
2026  3ed17d8ca68945e481112220b31f8439  25-34        married      iOS      U   
8011  f742ea750aeb4b13b4126ce956b5b5af  18-24         single  android      F   
4170  80732222e1484fe3a0589725c8f8e5d3  18-24         single      iOS      M   

       latitude   longitude         city state zip_code  frequency  amount  
3458  40.614660  -86.106432        Miami    IN    46959         13   345.0  
4139  45.408374 -122.691388  Lake Oswego    OR    97034         13   281.0  
1898  33.829743  -84.381529      Atlanta    GA    30312         11   280.0  
4581  33.792005  -84.442778      Atlanta    GA    30318         11   273.0  
7131  39.815990 -105.475322   Black Hawk    CO    80422         12   271.0  
...         ...         ...          ...   ...      ...        ...     ...  
7113  40.271322  -79.251145     Ligonier    PA    15658          1    21.0  
7148  33.794055  -84.377326      Atlanta    GA    30308          1    21.0  
2026  40.754782  -86.934599     Chalmers    IN    47929          1    21.0  
8011  33.794055  -84.377326      Atlanta    GA    30308          2    21.0  
4170  33.844371  -84.474050   Alpharetta    GA    30009          2    21.0  

[5291 rows x 12 columns]

The people who appear at the top of our dataframe `df` are the people who have given the highest funding to various projects. We can perform visualization on the top 500 (or 1000)

In [16]:
VizData = df.sort_values(by=['amount'], ascending=False)[:500]
VizData

session_id    age marital_status   device gender  \
3458  6b625dff333d480d9396cdc74f3c7275    55+        married      iOS      M   
4139  7f9c6efbf9964e8095772961442dd2d7  18-24         single  android      M   
1898  3ab197e44fc24289a08572bba665f702  35-44         single  android      F   
4581  8c9a458eafe7490eb7c3657d15e5fc10  18-24        married      iOS      F   
7131  db3131b5dce54211969002b2f554698b  18-24        married      iOS      F   
...                                ...    ...            ...      ...    ...   
4460  886b1f7ae0fe417f9efb8e5e8655185f  18-24         single  android      F   
6524  c7ba6a60842a4f82a7d2a8e337ddbd6f    55+         single  android      M   
4371  85be6d897b1a43c396d48120bc8c973c    55+         single  android      M   
6805  d069ee720fff4ef4b3192fcb8808fd4a  35-44        married  android      M   
2431  4afc877af793469b8aa2ba4bfdca64ae  18-24         single  android      F   

       latitude   longitude         city state zip_code  frequency  amount  
3458  40.614660  -86.106432        Miami    IN    46959         13   345.0  
4139  45.408374 -122.691388  Lake Oswego    OR    97034         13   281.0  
1898  33.829743  -84.381529      Atlanta    GA    30312         11   280.0  
4581  33.792005  -84.442778      Atlanta    GA    30318         11   273.0  
7131  39.815990 -105.475322   Black Hawk    CO    80422         12   271.0  
...         ...         ...          ...   ...      ...        ...     ...  
4460  33.752456  -84.392026      Atlanta    GA    30303          3   110.0  
6524  34.025111  -80.478151      Dalzell    SC    29040          6   110.0  
4371  29.552921  -95.666778   Sugar Land    TX    77479          3   110.0  
6805  48.299486 -116.553968    Sandpoint    ID    83864          6   110.0  
2431  33.724907  -84.468329      Atlanta    GA    30311          6   110.0  

[500 rows x 12 columns]

# Data Visualization:
We are finally ready to visualize our data `VizData`